employee-rostering-frontend/src/ui/components/DataTable.tsx

```diff
@@ -46,26 +49,25 @@ export abstract class DataTable<T, P extends DataTableProps<T>> extends React.Co
     this.state = {newRowData: null, currentFilter: (t) => true};
   }
 
-  abstract createNewDataInstance(): T;
   abstract displayDataRow(data: T): JSX.Element[];
-  abstract editDataRow(data: T): JSX.Element[];
-  abstract isValid(editedValue: T): boolean;
+  abstract editDataRow(data: ReadonlyPartial<T>, setProperty: (propertyName: keyof T, value: T[keyof T]|undefined) => void): JSX.Element[];
+  abstract isValid(editedValue: ReadonlyPartial<T>): editedValue is T;
 
   abstract updateData(data: T): void;
   abstract addData(data: T): void;
   abstract removeData(data: T): void;
 
   createNewRow() {
     if (this.state.newRowData === null) {
-      this.setState({...this.state, newRowData: this.createNewDataInstance()});
+      this.setState({...this.state, newRowData: {}});
```

Comment:
I think that spreading the state here is unnecessary as state updates are being merged https://reactjs.org/docs/state-and-lifecycle.html#state-updates-are-merged.